### DISCUSSION TITLE
Thoughts on py5 Documentation

### DISCUSSION THREAD

**Discussion by hx2A** [py5coding/py5generator#21] [General]
Here are my current thoughts on py5 documentation, both for the short term and long term. To be clear, I don't claim to have all the answers or a clear plan on what needs to be done. I'm hoping that with some discussion we can get a better idea of how to proceed in a way that is time-efficient and gets good results.

First, I think py5 documentation can be written with one of four sets of assumptions:

1 people who don't know python and haven't heard of Processing
2 people who don't know python but have experience with Processing or p5
3 people who know python but haven't heard of Processing
4 people who know python and have experience with Processing or p5

Right now, I believe there is a bottleneck that is making it difficult for other people to effectively use py5 or help out with any of the documentation, and that bottleneck is that too much information is only in my head and not in an accessible form for others. In the short term, most of the documentation I'm going to be writing is for people in group 4. I realize that group is smaller than the other groups and that a lot more that this needs to be created over the long term, but for the short term, I can provide interested contributors with the knowledge they need to help out with the other groups, and that is critical. I also know that I'm probably not the right person to write documentation for groups 1 or 2. Most likely writing documentation for groups 1 or 2 will require a longer period of time and a lot of people.

As to the tangible form of the content, I'll probably put notebooks in the [py5examples repo](https://github.com/hx2A/py5examples) so people can launch binder and see how everything works. I also believe that the [py5 website](http://py5.ixora.io/) I've put together is probably fine for the short term but will eventually need a redesign.

As to documentation organization, I thought this would be a good approach, and serves as a good long term goal:

https://labs.quansight.org/blog/2020/03/documentation-as-a-way-to-build-community/
https://documentation.divio.com/

Looking at other Processing projects like p5 or Processing, both libraries were created by a small number of people, then a community formed around them, with lots of people producing a wide array of content, ranging from videos to books to blog posts to tutorials. These libraries are accessible to all because of the collective efforts of lots of people. I'd like for the same to be true here.

Over the long term, I don't understand what I should be doing to best move things towards a good outcome. I do believe the answer includes being patient, focusing on developing a quality library, setting expectations about features and bugs, and good communication + listening.

@villares @tabreturn @BryanMed feel free to disagree with anything I've said here!

**Comment by tabreturn** [comment]
I think you've outlined it very well, @hx2A 

I would add that Processing owes a large part of its success to its group-1-appeal -- that is, creatives and first-time coders. I believe Python is the ideal language for those people (better than Java and JS). Even [Ben Fry thinks that "Python is terrific"](https://github.com/processing/processing4/wiki/Processing-4#goodbye-java). He goes on to say, "but it'd be awesome to have available NumPy and all those other things that make Python wonderful." Hello py5!

As a group-4-type, I'm happy to work through your documentation, adapting and curating it for group 1 & 2 types. As I've mentioned in another thread, it'll really help if we distribute something that runs with one click -- a bundled IDE (or one-click plugin) that won't require installing packages, runtimes, etc. Beginners and teachers don't want to wrestle with that stuff. The beginner documentation should match that IDE (think screenshots, instructions, etc.). Also, consider that beginners can't differentiate a library from a language from an IDE. "Processing" is just "Processing", IDE and all. "Python Mode" is a plugin, and nobody needs to know it's Jython+whatever. Can py5 be the name for a library and an IDE/plugin? Just something to think about.

I can also help out with design work. Good web design and branding can really 'sell' something. The fact that [processing.org](https://processing.org/) isn't responsive (side-scrolls in narrow windows) drives me mad; we can definitely do better than that, haha.

**Reply by hx2A** [reply]
> I would add that Processing owes a large part of its success to its group-1-appeal -- that is, creatives and first-time coders.  I believe Python is the ideal language for those people (better than Java and JS).

Excellent point, and I agree 100%. I do want py5 to be a good fit for people in group 1. And as @villares has said, have a low floor and high ceiling. I think it is important to provide a learning path that provides beginners with an avenue to achieve something that motivates them. For example, Scratch is good for beginners, but you can't really go very far with it, so people might not want to invest a lot of time and effort into learning it. Python and py5 can go very far, so beginners can see a path from their first steps to something more ambitious, such as a program that combines py5 with Python ML tools.

The py5 "ceiling" is building Python + Java hybrid applications. That's another important thing I need to document.

> Even [Ben Fry thinks that "Python is terrific"](https://github.com/processing/processing4/wiki/Processing-4#goodbye-java). He goes on to say, "but it'd be awesome to have available NumPy and all those other things that make Python wonderful." Hello py5!

Hmmm, I hadn't read that page before. I knew that Java 17 was planned at some point but didn't know it was 3 months away. BTW, py5 is using Jars for the alpha version of Processing 4. I'll need to make sure people know they need to change their JVM well in advance before the change.

If Ben Fry wants to rebuild the PDE and would like a HTML/CSS/JavaScript version of the PDE, he should consider using Jupyter. There are Jupyter kernels for Java. One could build a custom Java kernel that did the processing specific tasks. And the great thing about that is that you'd constantly get updates because of other people's contributions.

> As a group-4-type, I'm happy to work through your documentation, adapting and curating it for group 1 & 2 types. As I've mentioned in another thread, it'll really help if we distribute something that runs with one click -- a bundled IDE (or one-click plugin) that won't require installing packages, runtimes, etc. Beginners and teachers don't want to wrestle with that stuff. The beginner documentation should match that IDE (think screenshots, instructions, etc.).

I see your point. Having an easy to use distribution would make it easier to develop detailed documentation. Right now the installation process requires working in the terminal a little bit. I did get the Anaconda environment setup down to a single line, but even that won't be enough for everyone.

> Also, consider that beginners can't differentiate a library from a language from an IDE. "Processing" is just "Processing", IDE and all. "Python Mode" is a plugin, and nobody needs to know it's Jython+whatever. Can py5 be the name for a library and an IDE/plugin? Just something to think about.

You are right, and that isn't something I thought about much.

> I can also help out with design work. Good web design and branding can really 'sell' something. The fact that [processing.org](https://processing.org/) isn't responsive (side-scrolls in narrow windows) drives me mad; we can definitely do better than that, haha.

Well, one thing py5.ixora.io does well is it uses a better framework and looks great from mobile devices. I know the homepage is poor. I should look into improving that ASAP. Definitely, and web design work would be very helpful!!

Do you see https://readthedocs.org/ as a good potential place for documentation for this project?

**Reply by tabreturn** [reply]
> low floor and high ceiling ... Scratch is good for beginners ... people might not want to invest a lot of time and effort into learning it ... Python and py5 can go very far ...

All great thoughts; I agree.

> https://readthedocs.org/ as a good potential place for documentation for this project?

Interestingly, the p5 project is using this (https://p5.readthedocs.io/en/latest/). A lot of their documentation is just ported from the Processing website. I think readthedocs is fine to get started; from what I can tell, it's all markdown/rst, easy to shift elsewhere or adapt to another platform if it's ever necessary.

**Reply by hx2A** [reply]
readthedocs would be a great choice for py5's website documentation. Conveniently, all of the current documentation is already in rst format. I'm not using Sphinx though; for some reason I thought I would need to use Sphinx to generate the docs for that website.

I tried creating py5 documentation with Sphinx a while ago and the results were poor. I'm not very good at using that tool.

Crazy week, but I managed to move the ball forward with py5bot. I've encountered a few bugs related to the settings/setup feature I implemented a week or two ago. I'll need to get those fixed before doing the next release.

**Comment by villares** [comment]
Documentation & Feature request...
I'd like to  have some tips on how mortals can package/distribute a py5 project...

In my dreams, a killer py5 feature would be some "Export Application..." helper. I heard there is something like that for pygame, I'm not sure this is it: https://www.pygame.org/wiki/Pygame2exe

Windows installation hell and then Mac OS hell and Linux distros hell as well... I suppose we would need skillful person/people dedicated just for that. I would erect a giant statue* to those people. (I live in a very small apartment, so the giant statue may end up being a bit underwhelming).

**Reply by hx2A** [reply]
Packaging and distributing python packages normally is messy, but with py5's JVM and java path requirements, it will get even more complicated. Still, I'm optimistic it is at least possible to do this with pyinstaller. It might be a monster executable file, but it could work.

I added this to my todo list to give it a try and see how hard it is to get working.
